cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,300 @@
+    avgRunTime /= runs;
+
+    workflowRunRecords = sort(workflowRunRecords);
+
+    List<PercentileInformation> percentileInformationList = getPercentiles(workflowRunRecords, percentiles);
+
+    Map<String, ProgramRunDetails> actionToRunRecord = getActionRuns(workflowRunRecords);
+
+    Map<String, Map<String, String>> actionToStatistic = new HashMap<>();
+    for (Map.Entry<String, ProgramRunDetails> entry : actionToRunRecord.entrySet()) {
+      double avgForAction = 0;
+      for (long value : entry.getValue().getProgramRunList()) {
+        avgForAction += value;
+      }
+      avgForAction /= entry.getValue().getProgramRunList().size();
+      Map<String, String> actionMap = new HashMap<>();
+      actionMap.put("type", actionToRunRecord.get(entry.getKey()).getProgramType().toString());
+      actionMap.put("runs", Long.toString(entry.getValue().getProgramRunList().size()));
+      actionMap.put("avgRunTime", Double.toString(avgForAction));
+      actionToStatistic.put(entry.getKey(), actionMap);
+    }
+
+    for (Map.Entry<String, ProgramRunDetails> entry : actionToRunRecord.entrySet()) {
+      List<Long> runList = entry.getValue().getProgramRunList();
+      Collections.sort(runList);
+      for (double percentile : percentiles) {
+        long percentileValue = runList.get((int) ((percentile * runList.size()) / 100));
+        actionToStatistic.get(entry.getKey()).put(Double.toString(percentile), Long.toString(percentileValue));
+      }
+    }
+
+    return new WorkflowStatistics(startTime, endTime, runs, avgRunTime, percentileInformationList,
+                                  actionToStatistic);
+  }
+
+  private List<PercentileInformation> getPercentiles(List<WorkflowRunRecord> workflowRunRecords,
+                                                     List<Double> percentiles) {
+    int runs = workflowRunRecords.size();
+    List<PercentileInformation> percentileInformationList = new ArrayList<>();
+    for (double i : percentiles) {
+      List<String> percentileRun = new ArrayList<>();
+      int percentileStart = (int) ((i * runs) / 100);
+      for (int j = percentileStart; j < runs; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      percentileInformationList.add(
+        new PercentileInformation(i, workflowRunRecords.get(percentileStart).getTimeTaken(), percentileRun));
+    }
+    return percentileInformationList;
+  }
+
+  private List<WorkflowDataset.WorkflowRunRecord> sort(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
+    Collections.sort(workflowRunRecords, new Comparator<WorkflowRunRecord>() {
+      @Override
+      public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.WorkflowRunRecord o2) {
+        return Longs.compare(o1.getTimeTaken(), o2.getTimeTaken());
+      }
+    });
+    return workflowRunRecords;
+  }
+
+  private Map<String, ProgramRunDetails> getActionRuns(List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords) {
```

Comment:
rename this to getProgramRuns ? also we talked just returning List<ProgramRunDetails> as ``ProgramRunDetails`` has getProgramType, we do not need the map right ?